packages/sourcecred/src/plugins/github/fetchGithubRepo.js

```diff
@@ -328,10 +327,20 @@ export async function postQuery(
           console.error(error.error);
           break;
         case "GRAPHQL_ERROR":
-          console.error(
-            "Unexpected GraphQL error; this may be a bug in SourceCred: ",
-            JSON.stringify({postBody: postBody, error: error.error})
-          );
+          error.error.errors.forEach((error) => {
```

Comment:
This introduces a new assumption that error.error will be defined and will have an errors property, and this code will now crash if it doesn't--it's not null-safe. Looking at the [graphql spec](http://spec.graphql.org/June2018/#sec-Errors) it seems like this is probably an ok assumption to make, but it's still a little nervous-making and it's the kind of thing that'd be worth noting and justifying in the PR description in the future. I don't know enough about graphql and github's wrapping of it to feel comfortable counting on much of anything. I think I will ask that you add some sort of fallback strategy to make this nullish safe in the case that the API returns something unexpected.

Reply:
So I'm still learning about using javascript in a node context, and what flow type checking does and does not protect me from, so I did some reading and experimentation. I may be totally wrong, but here's what I seem to have figured out:

the 'error' that is being evaluated is generated between lines 156 and 193:

        if (x.errors) {
          if (
            x.errors.length === 1 &&
            x.errors[0].message.includes("it could be a GitHub bug")
          ) {
            return Promise.reject(
              ({
                type: "GITHUB_INTERNAL_EXECUTION_ERROR",
                error: x,
              }: GithubResponseError)
            );
          } else if (
            x.errors.length === 1 &&
            x.errors[0].type === "RATE_LIMITED"
          ) {
            return Promise.reject(
              ({type: "RATE_LIMIT_EXCEEDED", error: x}: GithubResponseError)
            );
          } else {
            return Promise.reject(
              ({type: "GRAPHQL_ERROR", error: x}: GithubResponseError)
            );
          }
        }
        if (x.data === undefined) {
          if (x.message && x.message.includes("Bad credentials")) {
            return Promise.reject(
              ({type: "BAD_CREDENTIALS", error: x}: GithubResponseError)
            );
          } else {
            // See https://github.com/sourcecred/sourcecred/issues/350
            return Promise.reject(
              ({type: "NO_DATA", error: x}: GithubResponseError)
            );
          }
        }
        return Promise.resolve(x.data);
      }),

It is a wrapper for the response error, and it checks to make sure the response error has an 'errors' before returning the wrapper. So we already know that if something has a type of 'GRAPHQL_ERROR', the error.error in that object has an errors. I suppose we don't know if error.error.errors is actually a list, but this would already have puked if it is not:

if (x.errors) {
          if (
            x.errors.length === 1 &&
            x.errors[0].message.includes("it could be a GitHub bug")

We also don't know if the list that is error.error.errors is a list of things that all have a 'type' attribute, but based on some experiments I've done, this code will not puke if 'type' is simply missing. It will just evaluate false on the ===:

if (error.type === "NOT_FOUND") {
              console.error(
                "Unable to find the specified repository. Please check for typos " +
                  "in the repository owner and name and confirm that you are using the correct token in " +
                  "$SOURCECRED_GITHUB_TOKEN"
              );
            } else {
              console.error(
                "Unexpected GraphQL error: " +
                  JSON.stringify({postBody: postBody, error: error})
              );
            }
          });

Does this evaluation seem correct? Do I have a bad assumption somewhere obvious(ish) here?

Reply:
Yeah, good research! The only assumption being made is that a known-truthy errors attribute is an array, which seems safe enough to assume. No change needed.